### ISSUE TITLE
Gradient Boosted Trees training does not use more of the CPU cores

### ISSUE THREAD

**Issue by gsaxena888** [haifengl/smile#623]
I'm using GBT for both classification and regression; on a large dataset, it does not appear to use more than 3 or 4 cores, instead of my full 32, during training. Is this the expected behaviour?

The reason I ask: XGBoost seems to imply that they can use GPUs or even distributed clusters (eg [xgboost on Kubernetes](https://xgboost.readthedocs.io/en/latest/tutorials/kubernetes.html)), so I would assume that it can use >>80% of CPUs on a single large multithreaded linux server. If you agree, are there any plans (or current ways) to make GBT on SMILE to use >>90 % of CPU cores for GBT training? (For prediction, single-threaded is fine.)

**Comment by haifengl** [comment]
Your data probably has very few columns.

**Comment by gsaxena888** [comment]
It has about 30 columns but 50,000 rows. In that case, does the fact that smile's GBT is only using 3 or 4 cores make sense? Is that a fundamental limitation of how gbt algoirhtms work? ie XGBoost (and other algorithms such as catboost, lightgbm) would also all not be able to make much use of dozens of CPU cores if all I have is about ~30 columns?

**Comment by haifengl** [comment]
We should use more cores. How do you check how many cores are used? Can you also check with `jps` how many threads are running?

BTW, Can you share the training parameters? If you can share your data, I can dive deeper. Thanks.

**Comment by gsaxena888** [comment]
Yes, I can share the data. Is there a format you think is best? (I was thinking of exporting the dataframe as an arrow file.) Also, in case it helps, I noticed that when running the "importance" method and also the "shap" routines that only a very small number of fields (eg about 3 or so columns) appear to be useful to predict the final output field value. (In fact, one field has an importance value that is 10x greater than the 2nd most important field.) Could that explain why only a few cores are being used? (Just guessing.)  Here are the parameters I use when running the gbt for regression analysis:
```
prop.setProperty("smile.gbt.trees", "18500"); //yes, that eighteen thousand five hundred
prop.setProperty( "smile.gbt.loss", "LeastSquares");
prop.setProperty("smile.gbt.node.size", "5");
prop.setProperty( "smile.gbt.shrinkage", "0.003");
prop.setProperty("smile.gbt.sample.rate", "0.5");
```

I also noticed a similar problem (ie only a few cores used) when I ran gbt for classification. (Again, only a small subset of columns, eg ~10 in this case, seem to be of "high importance" according to the output of the "importance" method.)

**Comment by haifengl** [comment]
csv file is good.

**Comment by haifengl** [comment]
I guess that your data don't have sufficient variance on many columns. So only 3 or so columns are inspected when building the tree. It is why only a few threads are busy.

**Comment by gsaxena888** [comment]
Here's the csv file (gzipped of course):
[rtModel.csv.gz](https://github.com/haifengl/smile/files/5598277/rtModel.csv.gz)

(The column that's we're trying to predict is called "Observed Rt" and the parameters used for the search were shared earlier in this post).

Also, I (partially) take back what I said. Although some columns are clearly not useful in this dataset (since they're all the same value), most of the columns (>>20) have (decent) variance. Any thoughts, therefore, on why the CPU usage is low during training? (eg only 3 or so threads are being used).

Also, if it is somehow related to the number of "useful" columns (which I think is different from whether the columns have a lot of variance), is this limitations true for all GBT approaches? That is (to the extend that you know of course), would XGBoost also be similarly limited to only use 3 or so cores? (I'm assuming that with GBT it is not possible to parallelize along the rows in some manner?)

**Comment by gsaxena888** [comment]
Any thoughts, @haifengl ,  on the attachment sent in the previous post?

**Comment by haifengl** [comment]
Most columns have close to 0 correlation with the dependent variable. Therefore, they are rarely used in the tree building.
